src/qibolab/_core/instruments/emulator/hamiltonians.py

```diff
@@ -0,0 +1,112 @@
+
+
+class Qubit(Config):
+    """Hamiltonian parameters for single qubit."""
+
+    frequency: float = 0
+    """Qubit frequency for 0->1."""
+    t1: float = 0
+    """Relaxation time."""
+    t2: float = 0
+    """Coherence time."""
+
+    @property
+    def operator(self):
+        """Time independent operator."""
+        return -np.pi * self.frequency * HZ_TO_GHZ * SIGMAZ
+
+    @property
+    def t_phi(self):
+        """T_phi computed from T1 and T2."""
+        return 1 / (1 / self.t2 - 1 / self.t1 / 2)
+
+    @property
+    def decoherence(self):
+        """Decoherence operator."""
+        assert self.t1 > 0 and self.t2 > 0
+        return np.sqrt(1 / self.t1) * L1 + np.sqrt(1 / self.t_phi / 2) * L2
+
+
+@dataclass
+class QubitDrive:
+    """Hamiltonian parameters for qubit drive."""
+
+    pulse: Pulse
+    """Drive pulse."""
+    frequency: float
+    """Drive frequency."""
+    sampling_rate: float = 1
+    """Sampling rate."""
+
+    @cached_property
+    def envelopes(self):
+        if isinstance(self.pulse, Delay):
+            return [np.zeros(len(self)), np.zeros(len(self))]
+        return self.pulse.envelopes(self.sampling_rate)
+
+    @cached_property
+    def operator(self):
+        """Time independent operator."""
+        return QUBIT_DRIVE
+
+    def __len__(self):
+        return int(self.pulse.duration)
+
+    def __call__(self, t, sample):
+        i, q = self.envelopes
+        if isinstance(self.pulse, Delay):
+            return 0
+        return self.pulse.amplitude * (
+            np.cos(2 * np.pi * self.frequency * t + self.pulse.relative_phase)
+            * i[sample]
+            + np.sin(2 * np.pi * self.frequency * t + self.pulse.relative_phase)
+            * q[sample]
+        )
+
+
+class HamiltonianConfig(Config):
+    """Hamiltonian configuration."""
+
+    kind: Literal["hamiltonian"] = "hamiltonian"
+    single_qubit: dict[str, Qubit] = Field(default_factory=dict)
+
+    @property
+    def hamiltonian(self):
+        return [qubit.operator for qubit in self.single_qubit.values()]
+
+    @property
+    def decoherence(self):
+        ops = []
+        for qubit in self.single_qubit.values():
+            if isinstance(qubit, list):
+                continue
+            else:
+                ops.append(qubit.decoherence)
+        return ops
+
+
+def waveform(pulse, channel, configs, updates=None) -> Optional[QubitDrive]:
+    """Convert pulse to hamiltonian."""
+    if updates is None:
+        updates = {}
+    # mapping IqConfig -> QubitDrive
+    if isinstance(configs[channel], IqConfig):
+        if channel in updates:
+            config = configs[channel].model_copy(update=updates[channel])
+            frequency = config.frequency
+        else:
+            frequency = configs[channel].frequency
+        if pulse.id in updates:
+            pulse = pulse.model_copy(update=updates[pulse.id])
+        return QubitDrive(pulse=pulse, frequency=frequency * HZ_TO_GHZ)
```

Comment:
This is more explicit (you know immediately what is happening in one case, reading little), and saves you some indentation:
```suggestion
    if not isinstance(configs[channel], IqConfig):
        return None

    if channel in updates:
        config = configs[channel].model_copy(update=updates[channel])
        frequency = config.frequency
    else:
        frequency = configs[channel].frequency
    if pulse.id in updates:
        pulse = pulse.model_copy(update=updates[pulse.id])
    return QubitDrive(pulse=pulse, frequency=frequency * HZ_TO_GHZ)
```